src/realm/set.hpp

```diff
@@ -0,0 +1,455 @@
+        // than an interleaving erase instruction, so emitting the instruction ensures that last "write" wins.
+        this->insert_repl(repl, it.index(), value);
+    }
+
+    m_tree->insert(it.index(), value);
+    CollectionBase::m_obj.bump_content_version();
+    return it.index();
+}
+
+template <class T>
+void Set<T>::insert_any(Mixed value)
+{
+    if constexpr (std::is_same_v<T, Mixed>) {
+        insert(value);
+    }
+    else {
+        if (value.is_null()) {
+            insert_null();
+        }
+        else {
+            insert(value.get<typename util::RemoveOptional<T>::type>());
+        }
+    }
+}
+
+template <class T>
+size_t Set<T>::erase(T value)
+{
+    REALM_ASSERT_DEBUG(!update_if_needed());
+    this->ensure_writeable();
+
+    auto b = this->begin();
+    auto e = this->end();
+    auto it = std::lower_bound(b, e, value, SetElementLessThan<T>{});
+
+    if (it == e || !SetElementEquals<T>{}(*it, value)) {
+        return not_found;
+    }
+
+    if (Replication* repl = m_obj.get_replication()) {
+        this->erase_repl(repl, it.index(), value);
+    }
+    m_tree->erase(it.index());
+    CollectionBase::m_obj.bump_content_version();
+    return it.index();
+}
+
+template <class T>
+void Set<T>::erase_any(Mixed value)
+{
+    if constexpr (std::is_same_v<T, Mixed>) {
+        erase(value);
+    }
+    else {
+        if (value.is_null()) {
+            erase_null();
+        }
+        else {
+            erase(value.get<typename util::RemoveOptional<T>::type>());
+        }
+    }
+}
+
+template <class T>
+inline void Set<T>::insert_null()
+{
+    insert(BPlusTree<T>::default_value(this->m_nullable));
+}
+
+template <class T>
+inline void Set<T>::erase_null()
+{
+    erase(BPlusTree<T>::default_value(this->m_nullable));
+}
+
+template <class T>
+inline void Set<T>::clear()
+{
+    ensure_created();
+    update_if_needed();
+    this->ensure_writeable();
+    if (size() > 0) {
+        if (Replication* repl = this->m_obj.get_replication()) {
+            this->clear_repl(repl);
+        }
+        m_tree->clear();
+        m_obj.bump_content_version();
+    }
+}
+
+template <class T>
+inline Mixed Set<T>::min(size_t* return_ndx) const
+{
+    if (size() != 0) {
+        if (return_ndx) {
+            *return_ndx = 0;
+        }
+        return *begin();
+    }
+    else {
+        if (return_ndx) {
+            *return_ndx = not_found;
+        }
+        return Mixed{};
+    }
+}
+
+template <class T>
+inline Mixed Set<T>::max(size_t* return_ndx) const
+{
+    auto sz = size();
+    if (sz != 0) {
+        if (return_ndx) {
+            *return_ndx = sz - 1;
+        }
+        auto e = end();
+        --e;
+        return *e;
+    }
+    else {
+        if (return_ndx) {
+            *return_ndx = not_found;
+        }
+        return Mixed{};
+    }
+}
+
+template <class T>
+inline Mixed Set<T>::sum(size_t* return_cnt) const
+{
+    return SumHelper<T>::eval(*m_tree, return_cnt);
```

Comment:
Maybe more of an item for the mixed implementation, but I'm not sure that these helpers will work for sum/avg of sets of mixed yet?

Reply:
I'm sure that they won't, or at least it will yield different results compared to MongoDB. :-)

If it's OK, I would like to defer this until https://jira.mongodb.org/projects/RCORE/issues/RCORE-359 is resolved.